cw/util/trace_library.py

```diff
@@ -0,0 +1,231 @@
+    key: bytearray
+    x_pos: Optional[int] = 0  # X position of probe
+    y_pos: Optional[int] = 0  # Y position of probe
+    trace_id: Optional[int] = 0
+
+
+class TraceLibrary:
+    """ Class for writing/reading traces to a database.
+
+    Either a new database is created or traces are appended to the existing
+    one. Traces are first written into memory and the flushed into the
+    database after reaching a trace memory threshold.
+    """
+    def __init__(self, db_name, trace_threshold, wave_datatype = np.uint16,
+                 overwrite = False):
+        if overwrite:
+            trace_lib_file = Path(db_name + ".db")
+            trace_lib_file.unlink(missing_ok=True)
+        self.engine = db.create_engine("sqlite:///" + db_name + ".db")
+        db_utils.create_database(self.engine.url)
+        self.session = sessionmaker(self.engine)()
+        self.metadata = db.MetaData()
+        self.traces_table = db.Table(
+            "traces",
+            self.metadata,
+            db.Column("trace_id", db.Integer, primary_key=True,
+                      autoincrement=True),
+            db.Column("wave", db.LargeBinary),
+            db.Column("plaintext", db.LargeBinary),
+            db.Column("ciphertext", db.LargeBinary),
+            db.Column("key", db.LargeBinary),
+            db.Column("x_pos", db.Integer),
+            db.Column("y_pos", db.Integer),
+        )
+        self.metadata_table = db.Table(
+            "metadata",
+            self.metadata,
+            db.Column("config", db.String),
+            db.Column("datetime", db.DateTime),
+            db.Column("bitstream_path", db.String),
+            db.Column("binary_path", db.String),
+            db.Column("bitstream", db.LargeBinary),
+            db.Column("binary", db.LargeBinary),
+            db.Column("offset", db.Integer),
+            db.Column("sample_rate", db.Integer),
+            db.Column("scope_gain", db.Float),
+            db.Column("trigger_level", db.Float),
+            db.Column("time_div", db.Float),
+            db.Column("ot_sca_commit", db.String),
+            db.Column("notes", db.String)
+        )
+        self.metadata.create_all(self.engine)
+        self.trace_mem = []
+        self.trace_mem_thr = trace_threshold
+        self.settingsDict = {}
+        self.wave_datatype = wave_datatype
+
+    def flush_to_disk(self):
+        """ Writes traces from memory into database.
+        """
+        if self.trace_mem:
+            query = db.insert(self.traces_table)
+            traces = []
+            for trace in self.trace_mem:
+                tr = asdict(trace)
+                del tr["trace_id"]
+                traces.append(tr)
+            self.session.execute(query, traces)
+            self.session.commit()
+            self.trace_mem = []
+
+    def write_to_buffer(self, trace):
+        """ Write traces into memory or into storage.
+
+        The trace first is written into memory. When reaching the trace
```

Comment:
Grammar nit: I think we probably want "The trace is first written ...". Also, the following sentence sounds like the traces are reaching the threshold. Maybe something like "When the length of the buffer reaches the memory threshold..."?